Makefile

```diff
@@ -1,4 +1,151 @@
-BASE_DIR ?= $(CURDIR)
+# Copyright 2021 TriggerMesh Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+KREPO      = triggermesh
+KREPO_DESC = TriggerMesh Open Source Components (sources and targets)
+
+
+BASE_DIR          ?= $(CURDIR)
+OUTPUT_DIR        ?= $(BASE_DIR)/_output
+
+# Dynamically generate the list of commands based on the directory name cited in the cmd directory
+COMMANDS          := $(shell ls $(BASE_DIR)/cmd)
+TARGETS           ?= linux/amd64
+
+BIN_OUTPUT_DIR    ?= $(OUTPUT_DIR)
+TEST_OUTPUT_DIR   ?= $(OUTPUT_DIR)
+COVER_OUTPUT_DIR  ?= $(OUTPUT_DIR)
+DIST_DIR          ?= $(OUTPUT_DIR)
+
+# Docker build variables
+DOCKER            ?= docker
+IMAGE_REPO        ?= gcr.io/triggermesh
+IMAGE_TAG         ?= latest
+IMAGE_SHA         ?= $(shell git rev-parse HEAD)
+
+# Rely on ko for dev style deployment
+KO                ?= ko
+
+# Go build variables
+GO                ?= go
+GOFMT             ?= gofmt
+GOLINT            ?= golangci-lint run --timeout 5m
+GOTOOL            ?= go tool
+GOTEST            ?= gotestsum --junitfile $(TEST_OUTPUT_DIR)/$(KREPO)-unit-tests.xml --format pkgname-and-test-fails --
+
+GOPKGS             = ./cmd/... ./pkg/apis/... ./pkg/function/... ./pkg/routing/... ./pkg/sources/... ./pkg/targets/... ./pkg/transformation/...
+LDFLAGS            = -extldflags=-static -w -s
+
+HAS_GOTESTSUM     := $(shell command -v gotestsum;)
+HAS_GOLANGCI_LINT := $(shell command -v golangci-lint;)
+
+.PHONY: help mod-download build install release test coverage lint fmt fmt-test images cloudbuild-test cloudbuild clean install-gotestsum install-golangci-lintall deploy undeploy verify
+
+all: codegen build test lint
+
+# Verify lint and tests
+install-gotestsum:
+ifndef HAS_GOTESTSUM
+	curl -SL https://github.com/gotestyourself/gotestsum/releases/download/v0.4.2/gotestsum_0.4.2_linux_amd64.tar.gz | tar -C $(shell go env GOPATH)/bin -zxf -
+endif
+
+install-golangci-lint:
+ifndef HAS_GOLANGCI_LINT
+	curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b $(shell go env GOPATH)/bin v1.26.0
+endif
+
+# Go pkg management
+mod-download: ## Download go modules
```

Comment:
Go downloads anything necessary when you compile, so do we really need that?

Reply:
Have been wondering about that for a while. Especially with the newer versions of go.